### PR TITLE
Microsoft Graph Device Management Integration

### PR DESCRIPTION
**Pull request by guyfreund**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/19628

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+
+''' CLIENT '''
+
+
+class MsGraphClient:
+    def __init__(self, ms_client):
+        self.ms_client = ms_client
+
+    def make_request(self, method, url_suffix, resp_type='json', data=None):
```
Comment (Itay4) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py]: why do we need this function? just call `_http_request`

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+import json
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+
+''' CLIENT '''
+
+
+class MsGraphClient:
```
Comment (Itay4) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py]: make sure to have request function in here

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    url: str = params.get('url', '')
+    server: str = url[:-1] if (url and url.endswith('/')) else url
+    # service base URL
+    base_url: str = server + '/v1.0'
+    app_name: str = 'ms-graph-device-management'
+    ok_codes: tuple = (200, 201, 202, 204)
+    use_ssl: bool = not params.get('insecure', False)
+    proxy: dict = handle_proxy()
+
+    if self_deployed:
```
Comment (Itay4) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py]: not sure why we didnt implement it in the module, lets move this condition to the module and make sure graphmail and singleuser support it

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.yml [hunk 4]
```diff
@@ -0,0 +1,821 @@
...
+    - contextPath: MSGraphDeviceManagement.Device.ConfigurationManagerClientEnabledFeatures.compliancePolicy
+      description: Whether compliance policy is managed by Intune
+      type: Boolean
+    - contextPath: MSGraphDeviceManagement.Device.ConfigurationManagerClientEnabledFeatures.windowsUpdateForBusiness
+      description: Whether Windows Update for Business is managed by Intune
+      type: Boolean
+    - contextPath: MSGraphDeviceManagement.Device.WiFiMacAddress
+      description: Wi-Fi MAC
+      type: String
+    - contextPath: MSGraphDeviceManagement.Device.HealthAttestationState
```
Comment (Arsenikr) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.yml]: please remove this, as this is a parent for other object and not an output itself

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.yml [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
```
Comment (Arsenikr) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.yml]: please remove

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+        index: int = HEADERS['raw_device'].index(header)
+        device_human_readable[HEADERS['device'][index]] = device.get(header)
+    return assign_params(**device_human_readable)
+
+
+''' COMMANDS '''
+
+
+def list_managed_devices_command(client: MsGraphClient, args: dict) -> None:
+    list_raw_devices, raw_response = client.list_managed_devices(args)
```
Comment (ShahafBenYakir) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py]: As we discussed, extract the relevant arguments from `args` and pass them explicitly to the client function.
Go over all commands and do the same.

File: Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+def get_managed_device_command(client: MsGraphClient, args: dict) -> None:
+    device_id: str = str(args.get('device_id'))
+    raw_response, device_id = client.get_managed_device(device_id)
+    device: dict = build_device_object(raw_response)
+    device_hr: dict = build_device_human_readable(raw_response)
+    entry_context: dict = {'MSGraphDeviceManagement.Device(val.ID === obj.ID)': device}
+    device_name: str = device.get('Name', '')
+    human_readable: str = f'Managed device {device_id} not found.'
+    if device:
+        human_readable = tableToMarkdown(f'Managed device {device_name}', device_hr, headers=HEADERS['device'])
```
Comment (ShahafBenYakir) [inline Packs/MicrosoftGraphDeviceManagement/Integrations/MicrosoftGraphDeviceManagement/MicrosoftGraphDeviceManagement.py]:
```suggestion

SPECIAL_HEADERS: dict =  {
    'raw_device': {
              'id': 'ID',
              'userId': 'User ID',
              'osVersion': 'OS Version',
              'imei': 'IMEI',
              'meid': 'MEID'
    }
}
HEADERS: dict = {
    'device': ['ID', 'User ID', 'Device Name', 'Operating System', 'OS Version', 'Email Address',
               'Manufacturer', 'Model', 'IMEI', 'MEID']
}


def convert_header_name(header: str) -> dict:
    """
   bla bla
    """
    return SPECIAL_HEADERS.get('raw_device').get(header, pascalToSpace(header))

        human_readable = tableToMarkdown(f'Managed device {device_name}', device_hr, headerTransform=convert_header_name, headers=HEADERS['device'], removeNull=True)
```
And now no need for `build_device_human_readable`

File: Packs/MicrosoftGraphDeviceManagement/TestPlaybooks/MSGraph_DeviceManagement_Test.yml [hunk 8]
```diff
@@ -0,0 +1,279 @@
+elasticcommonfields: {}
```
Comment (ShahafBenYakir) [inline Packs/MicrosoftGraphDeviceManagement/TestPlaybooks/MSGraph_DeviceManagement_Test.yml]: Change filename to `playbook-Microsoft_Graph_DeviceManagement_Test.yml`

**Comment by guyfreund** [comment]
@kirbles19  - yml file is ready to be reviewed.

**Pull request closed without merging**